enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,908 @@
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
+  Machines
+- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
+  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
+- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
+  Machines such that I can test changes with a single replica before applying the change to all replicas
+- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
+  scale down my control plane before adding new Control Plane Machines with newer configuration, notably, my
+  environment does not have capacity to add additional Machines during updates
+
+### API Extensions
+
+We will introduce a new `ControlPlaneMachineSet` CRD to the `machine.openshift.io/v1beta1` API group. It will be based
+on the spec and status structures defined below.
+
+We are intending to add this to the `v1beta1` API group as it is heavily tied to the existing `Machine` and `MachineSet`
+APIs within the `v1beta1` group. If we were to ever evolve the `Machine` to a `v1` API, the changes in that API version
+bump may mean that we also need to make changes to the spec of the `ControlPlaneMachineSet` as well. Additionally,
+as we currently unsure how we will handle Cluster API support, we may down the line need to make a breaking change to
+allow this project to work with Cluster API.
+
+```go
+type ControlPlaneMachineSetSpec struct {
+    // Replicas defines how many Control Plane Machines should be
+    // created by this ControlPlaneMachineSet.
+    // This field is immutable and cannot be changed after cluster
+    // installation.
+    // The ControlPlaneMachineSet only operates with 3 or 5 node control planes,
+    // 3 and 5 are the only valid values for this field.
+    // +kubebuilder:validation:Enum:=3;5
+    // +kubebuilder:default:=3
+    // +kubebuilder:validation:Required
+    Replicas *int32 `json:"replicas"`
+
+    // Strategy defines how the ControlPlaneMachineSet will update
+    // Machines when it detects a change to the ProviderSpec.
+    // +kubebuilder:default:={type: RollingUpdate}
+    // +optional
+    Strategy ControlPlaneMachineSetStrategy `json:"strategy,omitempty"`
+
+    // FailureDomains is the list of failure domains (sometimes called
+    // availability zones) in which the ControlPlaneMachineSet should balance
+    // the Control Plane Machines.
+    // This will be injected into the ProviderSpec in the
+    // appropriate location for the particular provider.
+    // This field is optional on platforms that do not require placement
+    // information, eg OpenStack.
+    // +optional
+    FailureDomains []string `json:"failureDomains,omitempty"`
```

Comment:
Is this field mutable?

What type of balancing will be done on the failure domains? For example, if I start with 1 failure domain and 3 replicas, then I add 3 more failure domains, can I expect the control planes to be re-distributed across the 3 failure domains? Similarly, if I start with 3 failure domains and later reduce down to one failure domain, can I expect the control plane machines to be re-created so that they are all in the single failure domain?

If the user wants to add failure domains that were not selected at install-time, will we provide instructions on how the user is expected to create the network infrastructure to support the new failure domains? Presumably there will not be an in-cluster operator that will do it for them.

Reply:
> Is this field mutable?

Initially we weren't planning to make it immutable, but did want to restrict, on creation, that the list matches the existing failure domains within the control plane machines. 

> What type of balancing will be done on the failure domains? For example, if I start with 1 failure domain and 3 replicas, then I add 3 more failure domains, can I expect the control planes to be re-distributed across the 3 failure domains? Similarly, if I start with 3 failure domains and later reduce down to one failure domain, can I expect the control plane machines to be re-created so that they are all in the single failure domain?

We have a use case where in vSphere, currently zones doesn't exist as a concept, but in 4.11 the SPLAT team are looking to introduce this. We don't want to stop customers from migrating their no-zone setup to multizone, and in theory this controller should be able to do that.

> If the user wants to add failure domains that were not selected at install-time, will we provide instructions on how the user is expected to create the network infrastructure to support the new failure domains? Presumably there will not be an in-cluster operator that will do it for them.

In the case above, we are expecting users to be given instructions for setting up the zone infrastructure and then being able to expand into the zones from there. I imagine we will do something similar for other platforms should customers wish to do so, however, for now we aren't planning to allow vertical scaling, so adding extra failure domains won't do anything, changing one of the values would cause a redeploy/rolling update